File: Linked list/reverse_nodes_in_k-group.py
"""Given a linked list, reverse the nodes of a linked list k at 
a time and return its modified list.
k is a positive integer and is less than or equal to the length 
of the linked list. If the number of nodes is not a multiple of k 
then left-out nodes in the end should remain as it is."""


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        if not head and not head.next:
            return head
        st = []
        curr = head
        sam = ListNode(0)
        temp = sam
        while curr:
            for i in range(k):
                st.append(curr)
                curr = curr.next
                if not curr:
                    break
            if len(st) == k:
                while st:
                    temp.next = st.pop()
                    temp = temp.next
            else:
                while st:
                    temp.next = st.pop(0)
                    temp = temp.next
        temp.next = None
        return sam.next
            

File: Linked list/ReverseNodesink-Group.cpp
//Problem Link:- https://leetcode.com/problems/reverse-nodes-in-k-group/

#include<bits/stdc++.h>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        if (k == 1){
            return head;
        }
        deque<ListNode*> pq;
        int cnt = 0;
        ListNode* curr = head;
        ListNode* start = new ListNode(-1);
        ListNode* dummy = start;
        while (curr != NULL){
            for(int i=0;i<k;i++){
                pq.push_back(curr);
                curr = curr->next;
                if (curr == NULL){
                    break;
                }
            }
            if (pq.size() == k){
                while (!pq.empty()){
                    start->next = pq.back();
                    start = start->next;
                    pq.pop_back();
                }
            }
            else{
                while (!pq.empty()){
                    start->next = pq.front();
                    start = start->next;
                    pq.pop_front();
                }
            }
        }
        start->next = NULL;
        return dummy->next;
    }
};